src/telemetry/mod.rs

```diff
@@ -0,0 +1,148 @@
+pub(crate) mod metrics;
+
+use itertools::Itertools;
+use once_cell::sync::OnceCell;
+use opentelemetry::{
+    global,
+    sdk::{metrics::PushController, trace::Config, Resource},
+    util::tokio_interval_stream,
+    KeyValue,
+};
+use opentelemetry_otlp::WithExportConfig;
+use std::{collections::VecDeque, time::Duration};
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};
+use url::Url;
+
+const TELEM_SERVICE_NAME: &str = "temporal-core-sdk";
+const LOG_FILTER_ENV_VAR: &str = "TEMPORAL_TRACING_FILTER";
+static TRACING_INIT: OnceCell<Result<GlobalTracingDats, anyhow::Error>> = OnceCell::new();
+
+/// Telemetry configuration options
+#[derive(Debug, Clone)]
+pub struct TelemetryOptions {
+    /// The url of the OTel collector to export telemetry and metrics to. Lang SDK should also
+    /// export to this same collector. If unset, telemetry is not exported and tracing data goes
+    /// to the console instead.
+    pub otel_collector_url: Option<Url>,
+    /// A string in the [EnvFilter] format which specifies what tracing data is included in
+    /// telemetry or console output. May be overridden by the `TEMPORAL_TRACING_FILTER` env
+    /// variable.
+    pub tracing_filter: String,
+}
+
+impl Default for TelemetryOptions {
+    fn default() -> Self {
+        Self {
+            otel_collector_url: None,
+            tracing_filter: "temporal_sdk_core=INFO".to_string(),
+        }
+    }
+}
+
+/// Things that need to not be dropped while telemetry is ongoing
+#[derive(Default)]
+struct GlobalTracingDats {
+    metric_push_controller: Option<PushController>,
+}
+
+/// Initialize tracing subscribers and output. Core [crate::init] calls this, but it may be called
+/// separately so that tests may choose to initialize tracing differently. If this function is
+/// called more than once, subsequent calls do nothing.
+///
+/// See [TelemetryOptions] docs for more on configuration.
+pub(crate) fn telemetry_init(opts: &TelemetryOptions) -> Result<(), anyhow::Error> {
+    TRACING_INIT.get_or_init(|| {
+        let mut globaldat = GlobalTracingDats::default();
+
+        let filter_layer = EnvFilter::try_from_env(LOG_FILTER_ENV_VAR)
```

Comment:
I'm still not sure I'd put the env var here, it should only be used for tests IMHO.

Reply:
There's no harm and they can be much easier to deal with operationally sometimes when you need to quickly change something. Having the option is nice

Reply:
I'd leave that up to the user to map env vars to config, everybody has their own configuration management systems and I wouldn't enforce a specific pattern.
We don't use env vars anywhere else in the core / node SDKs so this is a bit of a special case.